### PR TITLE
Core/adding clear to modelpart

### PR DESCRIPTION
**Pull request by RiccardoRossi**
This function adds a "Clear" functionality to a model_part. 

exact behaviour is to be discussed in @KratosMultiphysics/technical-committee , for now when doing Clear on a given submodelparts it destructs all of the submodelparts and all of the meshes in the part on which the Clear is called

### CODE REVIEW

File: kratos/tests/test_model_part.py [hunk 1]
```diff
@@ -66,6 +66,17 @@ def test_model_part_sub_model_parts(self):
         self.assertEqual(model_part.NumberOfSubModelParts(), 0)
         #print (model_part)
```
Comment (roigcarlo) [inline kratos/tests/test_model_part.py]:
```suggestion
```

File: kratos/sources/model_part.cpp [hunk 2]
```diff
@@ -142,6 +142,17 @@ ModelPart::~ModelPart()
 //       delete mpVariablesList;
 }
 
+void ModelPart::Clear()
+{
+    //wipe all of the submodelparts
```
Comment (loumalouomega) [inline kratos/sources/model_part.cpp]:
```suggestion
    // Wipe all of the submodelparts
```

File: kratos/sources/model_part.cpp [hunk 3]
```diff
@@ -142,6 +142,17 @@ ModelPart::~ModelPart()
...
+    mSubModelParts.clear();
+    mMeshes.clear();
+    mTables.clear();
+
+    this->AssignFlags( Flags() ); 
+
```
Comment (loumalouomega) [inline kratos/sources/model_part.cpp:103]:
```suggestion
```

File: kratos/tests/test_model_part.py [hunk 4]
```diff
@@ -66,6 +66,17 @@ def test_model_part_sub_model_parts(self):
         self.assertEqual(model_part.NumberOfSubModelParts(), 0)
         #print (model_part)
 
+
+        ##test that the clear function works
```
Comment (loumalouomega) [inline kratos/tests/test_model_part.py]:
```suggestion
        ## Test that the clear function works
```

**Review by loumalouomega — COMMENTED** [review]
Adding some minor suggestions

**Review by roigcarlo — DISMISSED** [review]
:+1:

**Comment by RiccardoRossi** [comment]
I am not sure of what i should do with the communicator and with the ProcessInfo...hence i'll wait to merge until we discuss it

**Review by loumalouomega — DISMISSED** [review]
OK

File: kratos/tests/test_model_part.py [hunk 5]
```diff
@@ -64,7 +64,36 @@ def test_model_part_sub_model_parts(self):
...
+        self.assertEqual(model_part.GetSubModelPart("Outlet").NumberOfSubModelParts(), 0)
+
+        model_part.Set(SLAVE)
+        self.assertTrue(model_part.Is(SLAVE))
+ 
+        #here the 
+        model_part.Clear()
+ 
+        self.assertEqual(model_part.NumberOfSubModelParts(),0)
+        self.assertEqual(len(model_part.Nodes),0)
```
Comment (jcotela) [inline kratos/tests/test_model_part.py]: Are nodes created at any point in the test? What should we expect to happen in that case? If I clear a submodelpart, will it delete its nodes/elements from its parent(s)?

Comment (RiccardoRossi) [reply kratos/tests/test_model_part.py]: Nodes are stored in meshes, and i delete the meshes, so what i am deleting are the smart_ptrs to the nodes. they are definitely not deleted from the father...

having said this, it is true that i am not creating nodes here. I should modify the test to include nodes creation

Comment (jcotela) [reply kratos/tests/test_model_part.py]: Thanks, if you could do that it would be very useful as a minimal example of usage for the new features.

Comment (RiccardoRossi) [reply kratos/tests/test_model_part.py]: @jcotela  dones. Also i am now calling clear on the Communicator which i think is to be expected

**Comment by RiccardoRossi** [comment]
after discussing in the @KratosMultiphysics/technical-committee this are the conclusions:

1 - DataValueContainer of modelpart must be deleted in the clear

define two functions:

- Clear     -->  (on root) Deletes all entities BUT preserves 
                      - variable list
                      - buffer size
                      - ProcessInfo
- ClearAll      -->  (on root)is  Completely equivalent to creating new modelpart (but without invalidation pointers to the existing one)
                      - empty variable list
                      - buffer_size = 1
                      - new ProcessInfo

on submodelparts the two versions are the same and consistency is mantained for buffer_size and variable_list

**Comment by dbaumgaertner** [comment]
This is just a comment, that this funciton is very welcome

**Comment by dbaumgaertner** [comment]
Sorry, I accidentally closed this PR. But reopened it already again. Just wanted to comment, that this function is very welcome from the point of view of @KratosMultiphysics/optimization. Thanks for including it.

**Comment by philbucher** [comment]
@dbaumgaertner are you sure you can use this?
It deletes all Nodes, elements etc

**Comment by dbaumgaertner** [comment]
@philbucher Ah ok. Then I understood Riccardos comment wrong. Sorry for the noise.

**Comment by maceligueta** [comment]
In any case, it would be great to know what was the actual function that you were expecting, @dbaumgaertner .

**Comment by dbaumgaertner** [comment]
@maceligueta I was looking for a function which allows clearing all results/modifications of an analysis on the model part. The goal would be, for example, to run different analyses with modified node coordinates but without rereading the whole case every time and while being sure that the results do not influence each other.

**Comment by pooyan-dadvand** [comment]
This sounds as `ClearData`

**Comment by philbucher** [comment]
is this still open?

**Comment by roigcarlo** [comment]
Yep we are still talking about this

**Comment by philbucher** [comment]
ping is this still active?

**Comment by pooyan-dadvand** [comment]
After discussing it again in @KratosMultiphysics/technical-committee we arrived to a new alternative which would be (hopefully) less confusing:

1. ModelPart::Clear --> Deletes all entities BUT preserves
- variable list
- buffer size
- ProcessInfo
Note that this removes all the sub model parts as well as nodes, elements, properties, etc.

2. Model::Reset(ModelPartName) reconstruct it to a **newly empty** modelpart. The pointers to that model part will still remain valid but everything is wiped out like the construction with default constructor.

**Comment by philbucher** [comment]
@KratosMultiphysics/technical-committee ping @RiccardoRossi

**Comment by pooyan-dadvand** [comment]
@sunethwarna will apply the changes conforming with the last comment

**Comment by rubenzorrilla** [comment]
> After discussing it again in @KratosMultiphysics/technical-committee we arrived to a new alternative which would be (hopefully) less confusing:
> 
>     1. ModelPart::Clear --> Deletes all entities BUT preserves
> 
> 
>     * variable list
> 
>     * buffer size
> 
>     * ProcessInfo
>       Note that this removes all the sub model parts as well as nodes, elements, properties, etc.
> 
> 
>     1. Model::Reset(ModelPartName) reconstruct it to a **newly empty** modelpart. The pointers to that model part will still remain valid but everything is wiped out like the construction with default constructor.

@sunethwarna will take care of implementing this.

**Comment by sunethwarna** [comment]
@rubenzorrilla @roigcarlo Could you have a look?

**Review by rubenzorrilla — COMMENTED** [review]
I think the `Clear` is correct but you didn't implement the `Reset`right.

**Review by pooyan-dadvand — APPROVED** [review]
@KratosMultiphysics/technical-committee approves it.

Thanks to @sunethwarna for the implementation

**Pull request merged**